Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2Connections.swift

```diff
@@ -0,0 +1,446 @@
+        ///            connection list.
+        func cleanup(_ context: inout CleanupContext) -> CleanupAction {
+            switch self.state {
+            case .starting:
+                return .keepConnection
+            case .backingOff:
+                context.connectBackoff.append(self.connectionID)
+                return .removeConnection
+            case .active(let connection, _, let usedStreams, _):
+                if usedStreams <= 0 {
+                    context.close.append(connection)
+                    return .removeConnection
+                } else {
+                    context.cancel.append(connection)
+                    return .keepConnection
+                }
+            case .draining(let connection, _, _):
+                context.cancel.append(connection)
+                return .keepConnection
+            case .closed:
+                preconditionFailure("Unexpected state for cleanup: Did not expect to have connections with this state in the state machine: \(self.state)")
+            }
+        }
+
+        init(connectionID: Connection.ID, eventLoop: EventLoop) {
+            self.connectionID = connectionID
+            self.eventLoop = eventLoop
+            self.state = .starting
+        }
+    }
+    
+    struct HTTP2Connections {
+        /// A connection's use. Did it serve in the pool or was it specialized for an `EventLoop`?
+        enum ConnectionUse {
+            case generalPurpose
+            case eventLoop(EventLoop)
+        }
+        
+        /// Information around an idle connection.
+        struct IdleConnectionContext {
+            /// The `EventLoop` the connection runs on.
+            var eventLoop: EventLoop
+            /// The connection's use. Either general purpose or for requests with `EventLoop`
+            /// requirements.
+            var use: ConnectionUse
+        }
+        
+        /// Information around the failed/closed connection.
+        struct FailedConnectionContext {
+            /// The eventLoop the connection ran on.
+            var eventLoop: EventLoop
+        }
+        
+        /// A connectionID generator.
+        private let generator: Connection.ID.Generator
+        /// The connections states
+        private var connections: [HTTP2ConnectionState]
+        
+        init(generator: Connection.ID.Generator) {
+            self.generator = generator
+            self.connections = []
+            self.connections.reserveCapacity(8) // just get's us some early breathing room
+        }
+        
+        // MARK: Connection creation
+        
+        struct AvailableConnectionContext {
+            var availableStreams: Int
+            var eventLoop: EventLoop
+        }
+        
+        mutating func createNewConnection(on eventLoop: EventLoop) -> Connection.ID {
+            // assert no connection exists on the requested eventLoop
+            assert(self.connections.first(where: { $0.eventLoop === eventLoop }) == nil)
```

Comment:
```suggestion
            assert(self.connections.allSatisfy(where: { $0.eventLoop !== eventLoop }))
```

There might be another catch here. If a connection is draining, we might want to create a new one on the same EL.